[clang-sycl-linker] Replace llvm-link with API calls

asudarsa opened the pull request:
This PR has the following changes:
1. Replace llvm-link with calls to linkInModule to link device files
2. Add -print-linked-module option to dump linked module for testing
3. Added a test to verify that linking is working as expected.

We will eventually move to using thin LTO for linking device inputs.

Thanks

asudarsa committed:
[clang-sycl-linker] Replace llvm-link with API calls

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
More changes

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
More changes

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
Add tests

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented on clang/test/Driver/clang-sycl-linker-test.cpp:
```diff
@@ -1,48 +1,35 @@
-// RUN: clang-sycl-linker --dry-run -triple spirv64 %t_1.bc %t_2.bc -o a.spv 2>&1 \
+// RUN: %clangxx -c -target spirv64 %s -o %t_1.o
+// RUN: %clangxx -c -target spirv64 %s -o %t_2.o
+// RUN: clang-sycl-linker --dry-run %t_1.o %t_2.o -o a.spv 2>&1 \
```
We do not need to pass -triple to clang-sycl-linker. We will get  that information from the device image. Also, the clang-sycl-linker can now accept .o fat object files (instead of .bc files). This is more aligned with the community changes in clang-linker-wrapper

jzc replied:
Doesn't `%clangxx -c -target spirv64 %s -o %t_1.o` create a SPIRV bytecode, not an (ELF) fat object?

asudarsa replied:
%clangxx -c '-target spirv64' will be used to run the front-end compiler (clang -cc1) for spirv64 target.
Here are the primary steps that will be run:
1.  clang -cc1 for device target
2. clang-offload-packager packages the device IR generated in step 2.
3. clang -cc1 for host target. packaged device code is embedded into the host object.
4. Additional clang calls to generate a fat object (.o file)

There is no backend compilation (or SPIR-V gen) for -c compilations.

Hope this answers your question.

Thanks

bader replied:
> Doesn't `%clangxx -c -target spirv64 %s -o %t_1.o` create a SPIRV bytecode, not an (ELF) fat object?

I think @jzc implies that the output file extensions should be `.spv` not `.o`.

asudarsa replied:
'clang++ -c' compilation will result in a .o (fat binary) file, not a .spv file.

Thanks

bader replied:
It will not unless you manually set the offloading model.
Your command should produce SPIR-V file.

asudarsa replied:
Hmm...Yes...you are right. It does create a SPIR-V file, albeit with a .o extension. Let me revisit this test.
Thanks

asudarsa replied:
Since these tests are all '--dry-run' tests (except the 'Unsupported file type' case, where I manually use packager to package the file), we do not reach the stage where we actually look inside the objects to see what it has.

However, it is too early to add such tests upstream. I have locally tested my changes with bitcodes and fat objects created using our downstream SYCL compiler.

I think I might as well just create dummy bitcode and object files here.

Thanks

asudarsa replied:
After some discussions, I have added test to verify the linked module.

Thanks

asudarsa commented on clang/test/Driver/clang-sycl-linker-test.cpp:
```diff
@@ -1,48 +1,35 @@
-// SIMPLE: "{{.*}}llvm-link{{.*}}" {{.*}}.bc {{.*}}.bc -o [[FIRSTLLVMLINKOUT:.*]].bc --suppress-warnings
-// SIMPLE-NEXT: "{{.*}}llvm-spirv{{.*}}" {{.*}}-o a.spv [[FIRSTLLVMLINKOUT]].bc
-//
-// Test that llvm-link is not called when only one input is present.
```
This cannot be tested in the current log that is generated via 'dry-run'.

asudarsa commented on clang/test/Driver/clang-sycl-linker-test.cpp:
```diff
@@ -1,48 +1,35 @@
-// DEVLIBS-NEXT: "{{.*}}llvm-link{{.*}}" -only-needed [[FIRSTLLVMLINKOUT]].bc {{.*}}lib1.bc {{.*}}lib2.bc -o [[SECONDLLVMLINKOUT:.*]].bc --suppress-warnings
-// DEVLIBS-NEXT: "{{.*}}llvm-spirv{{.*}}" {{.*}}-o a.spv [[SECONDLLVMLINKOUT]].bc
-//
-// Test a simple case with .o (fat object) as input.
```
Removed this test as .o objects are now supported.

asudarsa replied:
Outdated comment

jzc commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -179,60 +182,64 @@ Error executeCommands(StringRef ExecutablePath, ArrayRef<StringRef> Args) {
+      return createFileError(Filename, EC);
+    MemoryBufferRef Buffer = **BufferOrErr;
+    SmallVector<OffloadFile> Binaries;
+    if (Error Err = extractOffloadBinaries(Buffer, Binaries))
```
Doesn't this make it so `clang-sycl-linker` *only* accepts fat objects? And I thought clang-linker-wrapper extracted the device code from its fat object inputs and then passed them to the device linker directly, no wrapping in a fat object.

bader replied:
+1. I would expect this PR to keep `getInput` function untouched.

asudarsa replied:
Agree. +1 to keep getInput. We need extra logic to extract the LLVM IR Modules from the outputs of 'getInput' which currently returns list of input filenames.

Thanks

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -179,60 +182,64 @@ Error executeCommands(StringRef ExecutablePath, ArrayRef<StringRef> Args) {
+}
+
+/// Collect all fat objects from user input.
+Expected<SmallVector<std::string>> getObjectsFromArgs(const ArgList &Args) {
```
```suggestion
SmallVector<std::string> getInputs(const ArgList &Args) {
```
Drop `llvm::Expected` wrapper - function never returns an error.
`getInputs` better reflects the behavior of the function.

asudarsa replied:
I am dropping this entire function (getObjectFromArgs) and reverting back to getInputs. 

Thanks

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -179,60 +182,64 @@ Error executeCommands(StringRef ExecutablePath, ArrayRef<StringRef> Args) {
+  return DAL;
+}
+
+/// Collect all fat objects from user input.
```
The comment is misleading. Current implementation returns Args which are file names. I see no checks that the files are fat objects.

asudarsa replied:
I am getting rid of getBitcodeInputs as well. I am now using a new function: 'getBitcodeModule(std::string Filename, LLVMContext &C) to get the LLVM IR module from each file. The file can either be a fat object or a bitcode file.

Thanks

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -179,60 +182,64 @@ Error executeCommands(StringRef ExecutablePath, ArrayRef<StringRef> Args) {
+  for (Arg *A : Args)
+    DAL.append(A);
+  if (DryRun)
+    return DAL;
```
Why do we need to exit here in dry-run mode?

asudarsa replied:
I have removed 'getLinkerArgs' here. There is no reason to rely on reading the inputs to get the triple and arch. clang-linker-wrapper already does this and we just need to pass this information from there.

Thanks

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -179,60 +182,64 @@ Error executeCommands(StringRef ExecutablePath, ArrayRef<StringRef> Args) {
+  DAL.AddJoinedArg(nullptr, Tbl.getOption(OPT_arch),
+                   Args.MakeArgString(Input.front().getBinary()->getArch()));
+  DAL.AddJoinedArg(nullptr, Tbl.getOption(OPT_triple),
+                   Args.MakeArgString(Input.front().getBinary()->getTriple()));
```
Shouldn't arch and triple be the same for all input files?

asudarsa replied:
Yes.
As mentioned above, I am removing the 'getLinkerArgs' routine.

Thanks

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -179,60 +182,64 @@ Error executeCommands(StringRef ExecutablePath, ArrayRef<StringRef> Args) {
+/// Returns a new ArgList containg arguments used for the device linking phase.
+DerivedArgList getLinkerArgs(ArrayRef<OffloadFile> Input,
+                             const InputArgList &Args) {
+  DerivedArgList DAL = DerivedArgList(DerivedArgList(Args));
+  for (Arg *A : Args)
+    DAL.append(A);
```
Sorry, I don't understand these three lines. What is this code supposed to do?

asudarsa replied:
Sorry for the confusion. I have removed this routine. Also, this code had been borrowed from clang-linker-wrapper. It copies the original set of arguments and adds two new arguments for each target (-triple and -arch) on top of it.

Thanks

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -264,44 +271,86 @@ Expected<SmallVector<std::string>> getSYCLDeviceLibs(const ArgList &Args) {
-  CmdArgs.push_back("-o");
-  CmdArgs.push_back(*OutFileOrErr);
-  CmdArgs.push_back("--suppress-warnings");
-  if (Error Err = executeCommands(*LLVMLinkPath, CmdArgs))
-    return std::move(Err);
-  return *OutFileOrErr;
+  if (!BitcodeOutput)
+    return BitcodeOutput.takeError();
+  int FD = -1;
+  if (std::error_code EC = sys::fs::openFileForWrite(*BitcodeOutput, FD))
+    return errorCodeToError(EC);
```
Output file shouldn't be created in dry-run mode.

asudarsa replied:
Community code does create output file in dry-run mode (IIUC)
Example: 
https://github.com/llvm/llvm-project/blob/53b48301eb3bb4aba63678e9b5eb7cbea7745e34/clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp#L387

Thanks

bader replied:
> Community code does create output file in dry-run mode (IIUC)

I see that they create **temporary files**, not the output file. Temporary files are removed at end of the application work. 

Having said that, I don't think that creating temporary files is needed in the dry-run mode. `clang-linker-wrapper` do not run the tool using temporary file. So `clang-linker-wrapper` creates temporary file and removes it w/o using.

asudarsa replied:
I am also calling createTempFile here. I agree with you. We can attempt a cleanup of the dry-run support at some point of time.

Thanks

bader replied:
Actually, after thinking a bit more, I suggest that we remove dry-run mode from this function.
Left a separate comment about this. 
Please, mark this conversation as resolved.

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -264,44 +271,86 @@ Expected<SmallVector<std::string>> getSYCLDeviceLibs(const ArgList &Args) {
+  if (std::error_code EC = sys::fs::openFileForWrite(*BitcodeOutput, FD))
+    return errorCodeToError(EC);
+  if (DryRun || Verbose) {
+    std::for_each((*SYCLDeviceLibFilenames).begin(),
+                  (*SYCLDeviceLibFilenames).end(),
+                  [](std::string &Filename) { Filename += "; "; });
+    std::string LibInputs =
+        std::accumulate((*SYCLDeviceLibFilenames).begin(),
+                        (*SYCLDeviceLibFilenames).end(), std::string(""));
```
It can be done with a single std::accumulate call with something like:
```suggestion
    std::string LibInputs =
        std::accumulate(std::next((*SYCLDeviceLibFilenames).begin()),
                        (*SYCLDeviceLibFilenames).end(), (*SYCLDeviceLibFilenames).front(),
                        [](string a, string b) {
                          return std::move(a) + "; " + std::move(b);
                        });
```
Basically, use the overload with binary operator.

BTW, why not use `,` as files delimiter?

asudarsa replied:
That's a great suggestion. Thanks much. Will apply this.
Reason for using ';'
This seemed more readable: a;b;c;d; than a,b,c,d
I think, with your change, it might just be a,b,c,d

Will check that out.

Thanks

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -264,44 +271,86 @@ Expected<SmallVector<std::string>> getSYCLDeviceLibs(const ArgList &Args) {
+    auto FilenamesOrErr = getObjectsFromArgs(Args);
+    if (!FilenamesOrErr)
+      reportError(FilenamesOrErr.takeError());
+    std::for_each((*FilenamesOrErr).begin(), (*FilenamesOrErr).end(),
+                  [](std::string &Filename) { Filename += "; "; });
+    std::string Inputs = std::accumulate(
+        (*FilenamesOrErr).begin(), (*FilenamesOrErr).end(), std::string(""));
```
ditto

asudarsa replied:
Thanks

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -264,44 +271,86 @@ Expected<SmallVector<std::string>> getSYCLDeviceLibs(const ArgList &Args) {
+  for (auto &File : BitcodeInputFiles) {
+    auto TheImage = File.getBinary()->getImage();
+    std::unique_ptr<Module> ModOrErr =
+        parseIR(MemoryBufferRef(TheImage, ""), Err, C);
```
Can we use `getLazyIRModule` instead of `parseIR`?

asudarsa replied:
I tried that and ran into issues related to unique pointers. I will add a TODO and address this later.

Thanks

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -422,37 +470,39 @@ static Expected<StringRef> runLLVMToSPIRVTranslation(StringRef File,
-  // LLVM to SPIR-V translation step
-  auto SPVFile = runLLVMToSPIRVTranslation(*DeviceLinkedFile, Args);
+
+  LLVMContext C;
+
```
Unused. Please, move the declaration inside `linkDeviceCode` function.

asudarsa commented:
Hi @bader and @jzc 

Thanks for your kind feedback. I am refactoring my changes to provide an update. Please find my answers to your comments in place (adding them now).

Thanks

asudarsa committed:
Major refactoring to address review comments and also add back .bc support

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
Hi @bader and @jzc 

I have tried to address most of the comments in my recent commit. Please take a look when convenient. Sorry for major refactoring. I realized I was trying to 'unnecessarily' complicate things.

Thanks

asudarsa committed:
Minor changes

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -190,49 +191,48 @@ Expected<SmallVector<std::string>> getInput(const ArgList &Args) {
-    // This will be extended to support objects and SPIR-V IR files.
-    if (Magic != file_magic::bitcode)
-      return createStringError("Unsupported file type");
     BitcodeFiles.push_back(*Filename);
```
Please, add the check for input file format back. It's useful to detect errors.

asudarsa replied:
I have added the check back inside the getBitcodeModule function.

Thanks

bader replied:
Please, move it back here.

asudarsa replied:
Done. Thanks

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -190,49 +191,48 @@ Expected<SmallVector<std::string>> getInput(const ArgList &Args) {
+  if (M)
+    return std::move(M);
+
+  // Handle cases where input file is a fat object containing LLVM IR bitcode
+  // SYCL device libraries are provided as fat objects containing LLVM IR
+  // bitcode
```
Why SYCL device libraries are passed to the clang-sycl-linker via command line arguments? Doesn't the tool itself know the default location and default library names to link? I'm okay to let the user to customize these parameters via options, but I see that `clang-sycl-linker` can't link device libraries unless they are configured from outside.

If they are configured from outside, I would argue that device libraries must be LLVM bitcode format, not fat objects. AFAIK, they are distributed in the LLVM bitcode format. @mdtoguchi, am I right?

Anyway, processing two different sets of files in one function doesn't look like a good approach in this case. User inputs are processed by the code at the beginning of the function. Basically, we have two separate functions glued together into one. Instead of separating these two functions, I suggest using LLVM bitcode format for the device libraries.

asudarsa replied:
This function gives us the ability to handle both bitcode files and bitcode files embedded in fat objects. I would like to keep it like this to give some flexibility on how inputs are sent into clang-sycl-linker. We can do a cleanup once all pieces are put together.

Hope that's alright. I looked into the /install directory and I see .o .spv and .bc files. With wait for @mdtoguchi to chime in here about this.

Thanks

bader replied:
> This function gives us the ability to handle both bitcode files and bitcode files embedded in fat objects. I would like to keep it like this to give some flexibility on how inputs are sent into clang-sycl-linker. We can do a cleanup once all pieces are put together.

This doesn't make sense to me. User inputs should never be fat objects.

mdtoguchi replied:
AFAIK, there was agreement that the device libraries should be known by the tools and not required to be passed in from the driver or from the user calling `clang-linker-wrapper` directly.  I would also share Alexey's view that the device libraries should be .bc files.  We have full control over the device libraries so any potential flexibility given shouldn't be needed.

One interesting note though, current usage of `--offload-new-driver` will pass in fat objects for the device libraries instead of .bc files like we do for the old model.  I recall that @asudarsa made this distinction for some ease of use in the tool.

asudarsa replied:
Ok. I think we can use .bc files in new offload model as well. For now, I will continue to rely on the options for the device libraries to be passed in via options. Further discussions may be warranted on how we want to upstream the device library sources and build them in the upstream compiler.

Thanks

bader commented on clang/test/Driver/sycl-link-spirv-target.cpp:
```diff
@@ -4,6 +4,6 @@
 // RUN: touch %t.bc
 // RUN: %clangxx -### --target=spirv64 --sycl-link -Xlinker --llvm-spirv-path=/tmp \
-// RUN:   -Xlinker --library-path=/tmp -Xlinker --device-libs=lib1.bc,lib2.bc %t.bc 2>&1 \
+// RUN:   -Xlinker --library-path=/tmp -Xlinker --device-libs=lib1.o,lib2.o %t.bc 2>&1 \
```
What is the reason for changing the device library format?

asudarsa replied:
Reverted this. Thanks

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -264,44 +264,81 @@ Expected<SmallVector<std::string>> getSYCLDeviceLibs(const ArgList &Args) {
+Expected<StringRef> linkDeviceCode(ArrayRef<std::string> InputFiles,
+                                   const ArgList &Args) {
+  llvm::TimeTraceScope TimeScope("SYCL link device code");
+  LLVMContext C;
```
Please, move this declaration after the code handling the dry-run case.

bader commented on clang/test/Driver/clang-sycl-linker-test.cpp:
```diff
@@ -1,41 +1,43 @@
 // Tests the clang-sycl-linker tool.
 //
-// Test a simple case without arguments.
+// Test a simple case to link two fat objects.
```
The test doesn't use fat objects.
Please, remove this test case. The next test case does the same check. Instead, we should add a test checking the result of the linking user inputs.

asudarsa replied:
I had the same thought. Done in the latest commit.

bader replied:
Do you mean https://github.com/asudarsa/llvm-project/pull/4/commits/596488cdf625d2f2a84c381f0ddc4bc0968f73c0?
It does something different. It removes wrong test case, and it misses the check for the linker output.

asudarsa replied:
I updated the test and also added a new test to test the result

asudarsa committed:
Modify testcase and update CMakeLists.txt to make sure shared build works

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -264,44 +264,81 @@ Expected<SmallVector<std::string>> getSYCLDeviceLibs(const ArgList &Args) {
-      findProgram(Args, "llvm-link", {getMainExecutable("llvm-link")});
-  if (!LLVMLinkPath)
-    return LLVMLinkPath.takeError();
+  if (DryRun || Verbose) {
+    std::string Inputs =
+        std::accumulate(std::next(InputFiles.begin()), InputFiles.end(),
+                        InputFiles.front(), [](std::string a, std::string b) {
+                          return std::move(a) + ", " + std::move(b);
+                        });
+    std::string LibInputs = "";
+    if (!(*SYCLDeviceLibFiles).empty())
+      LibInputs = std::accumulate(
+          std::next((*SYCLDeviceLibFiles).begin()), (*SYCLDeviceLibFiles).end(),
+          (*SYCLDeviceLibFiles).front(), [](std::string a, std::string b) {
+            return std::move(a) + ", " + std::move(b);
+          });
+    errs() << formatv(
+        "sycl-device-link: inputs: {0} libfiles: {1} output: {2}\n", Inputs,
+        LibInputs, *BitcodeOutput);
+    if (DryRun)
+      return *BitcodeOutput;
+  }
```
According to my understanding, we should bailout in dry-run anymore as don't call the external tool. We should execute the function in any mode.
We probably should have some option to output the result of the linking to validate it.

asudarsa replied:
Added -print-linked-module option and also a test. Thanks

asudarsa committed:
Adding more tests and remove support for non-bitcode inputs

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
One more round of changes:
1. Removed early exit in linkDeviceCode for --dry-run
2. Removed reading module from fat binary
3. Added test to check linked output
4. Other minor fixes to get everything to work as expected.

Thanks again

asudarsa committed:
Pruned some tests

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
Minor fix

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
Minor fix

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -191,48 +193,26 @@ Expected<SmallVector<std::string>> getInput(const ArgList &Args) {
-  if (Error Err = executeCommands(*LLVMLinkPath, CmdArgs))
-    return std::move(Err);
-  return Args.MakeArgString(*OutFileOrErr);
+/// Handle cases where input file is a LLVM IR bitcode file
+/// When clang-sycl-linker is called via clang-linker-wrapper tool, input files
+/// are LLVM IR bitcode files
+// TODO: Support SPIR-V IR files
```
```suggestion
/// Handle cases where input file is a LLVM IR bitcode file.
/// When clang-sycl-linker is called via clang-linker-wrapper tool, input files
/// are LLVM IR bitcode files.
// TODO: Support SPIR-V IR files.
```

asudarsa replied:
Thanks. Will add this as part of my upcoming commit.

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -264,44 +244,82 @@ Expected<SmallVector<std::string>> getSYCLDeviceLibs(const ArgList &Args) {
+  if (!BitcodeOutput)
+    return BitcodeOutput.takeError();
+
+  // Get all SYCL device library files, if any
   auto SYCLDeviceLibFiles = getSYCLDeviceLibs(Args);
   if (!SYCLDeviceLibFiles)
     return SYCLDeviceLibFiles.takeError();
-  if ((*SYCLDeviceLibFiles).empty())
-    return InputFile;
 
-  Expected<std::string> LLVMLinkPath =
-      findProgram(Args, "llvm-link", {getMainExecutable("llvm-link")});
-  if (!LLVMLinkPath)
-    return LLVMLinkPath.takeError();
+  if (Verbose) {
```
This code is still organized for "early exit in dry-run mode". Let's align it with the function comment at the top to make it easier to read.

1. Move `if (Verbose)` to the end of the function just before the return statement.
2. Move BitcodeOutput declaration (lines 259-263) to line 317 (i.e. before writing the output to the file).
3. Move SYCLDeviceLibFiles declaration (lines 265-268) to line 301. I recommend using generic name for this variable like `DeviceLibs` because OpenMP offload will use them.

asudarsa replied:
Great reorg. Thanks.

bader commented on clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp:
```diff
@@ -264,44 +244,82 @@ Expected<SmallVector<std::string>> getSYCLDeviceLibs(const ArgList &Args) {
+  }
+
+  if (Args.hasArg(OPT_print_linked_module))
+    errs() << *LinkerOutput;
```
I don't understand the reason to print module to the error stream. LIT will fail is tests writes to the error stream. Redirecting the error stream to the output stream each for each test is annoying. Usually, you want your error stream to be empty. Non-empty error stream should signal about errors. I think other tools like clang driver print out verbose non-error output to the output stream.
We should discuss `clang-linker-wrapper` choice in the upstream and if `clang-sycl-linker` should follow the same approach.

```suggestion
    outs() << *LinkerOutput;
```

asudarsa replied:
Interesting. Will change it to outs(). Thanks

bader commented on clang/test/Driver/Inputs/libsycl.ll:
Let's put new files to clang/test/Driver/Inputs/SYCL/* directory.

bader commented on clang/test/Driver/Inputs/libsycl.ll:
```diff
@@ -0,0 +1,36 @@
+target triple = "spirv64"
+
+; Function Attrs: mustprogress nofree noinline norecurse nosync nounwind willreturn memory(none)
+define spir_func noundef range(i32 -2147483643, -2147483648) i32 @_Z9lib_func1i(i32 noundef %a) local_unnamed_addr #0 {
+entry:
+  %add = add nsw i32 %a, 5
+  ret i32 %add
+}
```
I would prefer to have simpler code for the library (and tests):
1. Unmingled names.
2. No unnecessary attributes.
3. Less code. Do we need 4 library functions? I would leave just one:

```llvm
define spir_func i32 addFive(i32 %x) {
  %res = add i32 %x, 5
  ret i32 %res
}
```

It's easier to maintain and read.

bader commented on clang/test/Driver/Inputs/libsycl.ll:
```diff
@@ -0,0 +1,36 @@
+; ModuleID = 'libsycl.cpp'
+source_filename = "libsycl.cpp"
+target datalayout = "e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-n8:16:32:64-G1"
+target triple = "spirv64"
```
```suggestion
target triple = "spirv64"
```
Please, remove from the tests as well.

asudarsa replied:
Done as part of upcoming commit. Thanks

bader commented on clang/test/Driver/Inputs/test1.ll:
```diff
@@ -0,0 +1,28 @@
+target triple = "spirv64"
+
+; Function Attrs: mustprogress noinline
+define spir_func noundef i32 @_Z9foo_func1ii(i32 noundef %a, i32 noundef %b) local_unnamed_addr #0 {
```
Please, use one notation for function prefixes and file names.
Either rename `test1.ll` to `foo.ll` and `test2.ll` to `bar.ll` or use `test1_` and `test2_` function name prefixes.

bader commented on clang/test/Driver/Inputs/test2.ll:
```diff
@@ -0,0 +1,22 @@
+}
+
+; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
+define spir_func noundef i32 @_Z9bar_func2iii(i32 noundef %c, i32 noundef %d, i32 noundef %e) local_unnamed_addr #1 {
```
Let's remove func2 and keep only one function in this test.

bader reviewed (commented):
@asudarsa, the code looks good in general. I have just a few style comments.
I would also make new tests smaller and simpler.

bader commented on clang/test/Driver/link-device-code.test:
```diff
@@ -0,0 +1,22 @@
+; CHECK-DEVICE-LIB: define {{.*}}foo_func1{{.*}}
+; CHECK-DEVICE-LIB: define {{.*}}foo_func2{{.*}}
+; CHECK-DEVICE-LIB: define {{.*}}bar_func1{{.*}}
+; CHECK-DEVICE-LIB: define {{.*}}bar_func2{{.*}}
```
CHECK-DEVICE-LIB doesn't check that device libraries are linked.
I would also check that CHECK-SIMPLE doesn't define library functions.

asudarsa replied:
Updating the tests now. 
Thanks

asudarsa committed:
Simplify tests and update comments

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
Added -dry-run option to testing

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

bader commented on a.spv:
Please, remove this file.

asudarsa replied:
Ah. I deleted that specifically before merge

bader reviewed (commented):
LGTM.

asudarsa committed:
Remove unnecessary file

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa commented:
Removed the file. Thanks @bader for the great feedback. I learnt a lot and will try my incorporate these learnings in upcoming patches.

Will prepare and submit an upstream patch in a bit.

asudarsa committed:
Valid LLVM IR bitcode input files are required

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

asudarsa committed:
Minor changes to align with changes in SPIR-V backend PR

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>
